rfc/router-storage.md

```diff
@@ -0,0 +1,767 @@
+      { 
+        "out": {
+          "ok": {
+            // result of the add is the content (location) commitment
+            // that is produced as result of "bafy..accept"
+            "size": 2_097_152,
+            "address": {
+              "url": "https://node1.storage/upload/xxxa-presigned",    
+              "headers": {},
+              "expires": 1712903125
+            }
+          }
+        },
+        "next": []
+      },
+    ],
+  },
+}
+```
+
+Legacy receipt:
+
+```json
+{ 
+  // "/": "bafy..work",
+  "iss": "did:web:node1.storage",
+  "aud": "did:web:web3.storage",
+  "ran": {
+    "/": "bafyreia5tctxekbm5bmuf6tsvragyvjdiiceg5q6wghfjiqczcuevmdqcu"
+  },
+  "out": {
+    "ok": {
+      "size": 2_097_152,
+      "address": {
+        "url": "https://node1.storage/upload/xxxa-presigned",    
+        "headers": {},
+        "expires": 1712903125
+      }
+    }
+  },
+  "fx": [],
+}
+```
+
+4. On success, router creates a success receipt for the original user
+
+Receipt:
+```json
+{ 
+  // "/": "bafy..work",
+  "iss": "did:web:web3.storage",
+  "aud": "did:web:web3.storage",
+  "cmd": "/ucan/assert",
+  "sub": "did:web:web3.storage",
+  "args": {
+    "assert": [
+      // refers to the invocation from the example
+      { "/": "bafy..allocate" },
+      // refers to the receipt corresponding to the above invocation
+      { 
+        "out": {
+          "ok": {
+            "size": 2_097_152,
+            "address": {
+              "url": "https://node1.storage/upload/xxxa-presigned",    
+              "headers": {},
+              "expires": 1712903125
+            }
+          }
+        },
+        "next": []
+      },
+    ],
+  },
+}
+```
+
+Legacy receipt:
+
+```json
+{ 
+  // "/": "bafy..work",
+  "iss": "did:web:web3.storage",
+  "aud": "did:web:web3.storage",
+  "ran": {
+    "/": "bafy..allocate"
+  },
+  "out": {
+    "ok": {
+      // result of the add is the content (location) commitment
+      // that is produced as result of "bafy..accept"
+      "size": 2_097_152,
+      "address": {
+        "url": "https://node1.storage/upload/xxxa-presigned",    
+        "headers": {},
+        "expires": 1712903125
+      }
+    }
+  },
+  "fx": [],
+}
+```
+
+The router should also store a BlobAllocate record:
+
+```json
+{
+  "did": "did:web:node1.storage",
+  "blob": {
+    // multihash of the blob as byte array
+    "digest": { "/": { "bytes": "mEi...sfKg" } },
+    // size of the blob in bytes
+    "size": 2_097_152,
+  },
+}
+```
+
+4. On fail, router adds the failed nodes to a list of "alreadyFailed", and returns to step 2.
+
+5. If all nodes fail, return a fail receipt to the user for `blob/allocate`
+
+#### Router handling of blob accept
+
+After the user puts their blob to the presigned URL, they signal to the router that `blob/accept` can proceed by sending them the UCAN receipt for `http/put`(currently this is handled differently between legacy and 1.0 UCAN)
```

Comment:
Is there a reason why do we need to route these as opposed to letting user deal directly with the storage node that allocated and let them perform their own `accept` ? I would expect that `blob/add` simply describes coordination with storage nodes by forking set of `allocation → put → accept` tasks flows and then combines their results into an output.

Reply:
I'm not going to comment on this section of the spec as I think we probably should not try to route `blob/accept` but rather combine outputs from storage nodes. It also would remove dependency on local state that otherwise requires side-effects and can be racy.

Reply:
how does the router know about the resulting location commitment then?

arguably I guess instead of keeping a table of blob allocations, it could just query our indexing system.

Reply:
Why does it need to know of it ? If it needs to we can add a step in the pipeline to that effect. That said I don't think router needs to know or care about it, if user wants to utilize received commitment it would have to authorize gateway by delegating commitment effectively giving it permission to perform reads at user expense. If it never shared with gateway or libp2p node user does not wish to share read permission and that seems very reasonable choice IMO.